server/bin/test_data_importer.py

```diff
@@ -0,0 +1,706 @@
+        ordered = []
+        processed_pids = []
+
+        def traverse_product(product):
+            if product['id'] in processed_pids:
+                return
+
+            if 'derived_product_id' in product:
+                if product['derived_product_id'] not in product_map:
+                    raise ReferenceError("product references a non-existent product: %s (%s) => %s" % (product['name'], product['id'], product['derived_product_id']))
+
+                traverse_product(product_map[product['derived_product_id']])
+
+            if 'provided_products' in product:
+                for pid in product['provided_products']:
+                    if pid not in product_map:
+                        raise ReferenceError("product references a non-existent product: %s (%s) => %s" % (product['name'], product['id'], pid))
+
+                    traverse_product(product_map[pid])
+
+            processed_pids.append(product['id'])
+            ordered.append((owner_key, product))
+
+        for product in product_map.values():
+            traverse_product(product)
+
+        return ordered
+
+    products = []
+
+    if hosted_test:
+        product_map = {}
+
+        # In hosted mode, we don't need per-org products, so we only add the global products once
+        if 'products' in data:
+            for product in data['products']:
+                product_map[product['id']] = product
+
+        if 'owners' in data:
+            for owner in data['owners']:
+                if 'products' in owner:
+                    for product in owner['products']:
+                        product_map[product['id']] = product
+
+        products.extend(order_products(None, product_map))
+    else:
+        # In standalone mode, we can only add stuff if we have an org to add it to
+        if 'owners' in data:
+            for owner in data['owners']:
+                owner_key = owner['name']
+                product_map = {}
+
+                if 'products' in data:
+                    for product in data['products']:
+                        product_map[product['id']] = product
+
+                if 'products' in owner:
+                    for product in owner['products']:
+                        product_map[product['id']] = product
+
+                products.extend(order_products(owner_key, product_map))
+
+    return products
+
+# Performs content mapping for a given product
+def map_content_to_product(owner_key, product_data):
+    content_map = {}
+
+    if 'content' in product_data:
+        # test data stores this as an array of arrays for reasons. Convert it to a dictionary and pass it
+        # to Candlepin
+        for content_ref in product_data['content']:
+            content_map[content_ref[0]] = content_ref[1]
+
+    if content_map:
+        cp.add_content_to_product(owner_key, product_data['id'], content_map)
+
+
+# Process a given JSON file, creating the test data it defines
+def process_file(cp, options, filename):
```

Comment:
This function should be refactored to several functions like `create_owners()`, `create_users()`, etc.

Reply:
done